corehq/apps/enterprise/enterprise.py

```diff
@@ -635,3 +634,153 @@ def rows_for_domain(self, domain_obj):
+        ]
+
+    @property
+    def rows(self):
+        rows = []
+        for domain in self.account.get_domains():
+            rows.extend(self.rows_for_domain(domain))
+        return rows
+
+    @property
+    def total(self):
+        total_out_of_date_rows = 0
+        total_last_builds = 0
+
+        for domain in self.account.get_domains():
+            domain_out_of_date_rows, domain_last_builds = self.total_for_domain(domain)
+            total_out_of_date_rows += domain_out_of_date_rows
+            total_last_builds += domain_last_builds
+
+        total_up_to_date = total_last_builds - total_out_of_date_rows
+
+        return _format_percentage_for_enterprise_tile(total_up_to_date, total_last_builds)
+
+    def rows_for_domain(self, domain):
+        rows = []
+        app_name_by_id = {}
+        apps = get_app_ids_in_domain(domain)
+
+        for user, build, latest_version in self._get_user_builds(domain, apps):
+            if is_out_of_date(str(build['build_version']), str(latest_version)):
+                app_id = build['app_id']
+                if app_id not in app_name_by_id:
+                    app_name_by_id[app_id] = Application.get_db().get(app_id).get('name')
+                rows.append([
+                    user['username'],
+                    domain,
+                    app_name_by_id[app_id],
+                    latest_version,
+                    build['build_version'],
+                    self.format_date(DateTimeProperty.deserialize(build['build_version_date'])),
+                ])
+
+        return rows
+
+    def total_for_domain(self, domain):
+        apps = get_app_ids_in_domain(domain)
+        total_last_builds = 0
+        total_out_of_date = 0
+
+        for user, build, latest_version in self._get_user_builds(domain, apps):
+            total_last_builds += 1
+            if is_out_of_date(str(build['build_version']), str(latest_version)):
+                total_out_of_date += 1
+
+        return total_out_of_date, total_last_builds
+
+    def _get_user_builds(self, domain, apps):
+        user_query = (UserES()
+            .domain(domain)
+            .mobile_users()
+            .source([
+                'username',
+                'reporting_metadata.last_builds',
+            ]))
+        for user in user_query.run().hits:
+            last_builds = user.get('reporting_metadata', {}).get('last_builds', [])
+            for build in last_builds:
+                app_id = build.get('app_id')
+                build_version = build.get('build_version')
+                if app_id not in apps or not build_version:
+                    continue
+                build_version_date = DateTimeProperty.deserialize(build.get('build_version_date'))
+                all_builds = self.get_app_builds(domain, app_id, limit=self.INITIAL_QUERY_LIMIT)
+                latest_version = self.get_latest_build_version_at_time(all_builds, build_version_date)
+                if not latest_version:
```

Comment:
It seems unfortunate that we have to fetch all builds to find the latest version available at the time of use. Is it possible to have a more targeted query using the built_on and/or last_released properties? The downside is that would probably be less effective for the current caching mechanism and in turn might result in more ES queries which could be worse, even if they are targeted. Is that something you have explored?

Reply:
A more targeted query takes 200 sec in total...I don't understand why it takes that long...
```
app_es = (
    AppES()
    .domain(domain)
    .is_build()
    .app_id(app_id)
    .sort('version', desc=True)
    .is_released()
    .source(['_id', 'version'])
    .filter(
        OR(
            range_filter('last_released', lte=before_date),
            range_filter('built_on', lte=before_date)
        )
    )
    .size(1)
)
```

Reply:
Jing and I offlined about this and I believe cleared this one up. Thanks for doing the exploration on this and sorry that it didn't help anything 😅